projects/opendr_ws/src/perception/scripts/skeleton_based_action_recognition.py

```diff
@@ -0,0 +1,205 @@
+        self.pose_estimator.download(path=".", verbose=True)
+        self.pose_estimator.load("openpose_default")
+
+        # Initialize the skeleton_based action recognition
+        if model == 'stgcn':
+            self.action_classifier = SpatioTemporalGCNLearner(device=device, dataset_name='nturgbd_cv',
+                                                              method_name=model)
+        elif model == 'pstgcn':
+            self.action_classifier = ProgressiveSpatioTemporalGCNLearner(device=device, dataset_name='nturgbd_cv',
+                                                                         topology=[5, 4, 5, 2, 3, 4, 3, 4])
+        model_saved_path = self.action_classifier.download(path="./pretrained_models/"+model,
+                                                           method_name=model, mode="pretrained",
+                                                           file_name=model+'_ntu_cv_lw_openpose')
+        self.action_classifier.load(model_saved_path, model+'_ntu_cv_lw_openpose')
+
+    def listen(self):
+        """
+        Start the node and begin processing input data
+        """
+        rospy.init_node('opendr_skeleton_based_action_recognition', anonymous=True)
+        rospy.loginfo("Skeleton-based action recognition node started!")
+        rospy.spin()
+
+    def callback(self, data):
+        """
+        Callback that process the input data and publishes to the corresponding topics
+        :param data: input message
+        :type data: sensor_msgs.msg.Image
+        """
+
+        # Convert sensor_msgs.msg.Image into OpenDR Image
+        image = self.bridge.from_ros_image(data, encoding='bgr8')
+
+        # Run pose estimation
+        poses = self.pose_estimator.infer(image)
+        if len(poses) > 2:
+            # select two poses with highest energy
+            poses = _select_2_poses(poses)
+
+        # Get an OpenCV image back
+        image = np.float32(image.numpy())
+        #  Annotate image and publish results
+        for pose in poses:
+            if self.pose_publisher is not None:
+                ros_pose = self.bridge.to_ros_pose(pose)
+                self.pose_publisher.publish(ros_pose)
+                # We get can the data back using self.bridge.from_ros_pose(ros_pose)
+                # e.g., opendr_pose = self.bridge.from_ros_pose(ros_pose)
+                draw(image, pose)
+
+        if self.image_publisher is not None:
+            message = self.bridge.to_ros_image(np.uint8(image), encoding='bgr8')
+            self.image_publisher.publish(message)
+
+        num_frames = 300
+        poses_list = []
+        for _ in range(num_frames):
+            poses_list.append(poses)
+        skeleton_seq = _pose2numpy(num_frames, poses_list)
+
+        # Run action recognition
+        category = self.action_classifier.infer(skeleton_seq)
+        category.confidence = category.confidence.max()
+        category.description = NTU60_ClASSES[category.data]
```

Comment:
Shouldn't this be part of the original learner implementation?

Reply:
You are right. The inference implementation of learners is now modified and this part is removed from the ROS node implementation.